addons/mail/static/src/models/thread/thread.js

```diff
@@ -1155,6 +1169,56 @@ function factory(dependencies) {
             return [['unlink']];
         }
 
+        /**
+         * @private
+         * @returns {mail.message|undefined}
+         */
+        _computeLastNonTransientMessage() {
+            const {
+                length: l,
+                [l - 1]: lastMessage,
+            } = this.orderedNonTransientMessages;
+            if (lastMessage) {
+                return [['link', lastMessage]];
+            }
+            return [['unlink']];
+        }
+
+        /**
+         * Adjusts the last seen message received from the server to consider
+         * the following messages also as read if they are either transient
+         * messages or messages from the current partner.
+         *
+         * @private
+         * @returns {integer}
+         */
+        _computeLastSeenByCurrentPartnerMessageId() {
+            const firstMessage = this.orderedMessages[0];
+            if (
+                firstMessage &&
+                this.lastSeenByCurrentPartnerMessageId &&
+                this.lastSeenByCurrentPartnerMessageId < firstMessage.id
+            ) {
+                // no deduction can be made if there is a gap
+                return this.lastSeenByCurrentPartnerMessageId;
+            }
+            let lastSeenByCurrentPartnerMessageId = this.lastSeenByCurrentPartnerMessageId;
+            for (const message of this.orderedMessages) {
+                if (message.id <= this.lastSeenByCurrentPartnerMessageId) {
+                    continue;
+                }
+                if (
+                    message.author === this.env.messaging.currentPartner ||
+                    message.isTransient
+                ) {
+                    lastSeenByCurrentPartnerMessageId = message.id;
+                    continue;
+                }
+                return lastSeenByCurrentPartnerMessageId;
+            }
+            return lastSeenByCurrentPartnerMessageId;
+        }
```

Comment:
Same remark. It should be a relation, not a compute attribute which is derived from the identifying field of another record.

Reply:
The target message is not necessarily known, `lastSeenByCurrentPartnerMessageId` is an existing field (with even an explanation as to why it is an integer), this compute is just making adaptations.

Reply:
> target message is not necessarily known

This doesn't make sense. If we are using message ids, we know the message, even if it's just the id.
Any code that solely relies on message ids (integers) must instead rely on messages (records).
If you have the message id, you necessarily have the message record (and if not, should have been inserted).

Reply:
From the look of it, initial workaround of using message ids as "date" motivates doing this, and it goes in a very bad direction for code maintenance: it introduce fields of a record that refers to attribute of another record that has special properties (use as date). There are many pitfalls with this approach: mixing ids & records, heavy relying on initial bad design (= makes harder to change it), use of sentinel values (0 & `undefined` _are_ sentinel values, can't deny and motivate these values with special meaning outside of actual message id, which is exactly what sentinel values are...).

Instead of using message ids as date, we should have a clean logical representation of something that is date. This piece of information is immutable and isn't affected by message being deleted and or moved to another thread.

Goes without saying clean way is master. This PR is 14.0 fix with already too many changes for my taste. It should not introduce bad stuff like the fields that are message ids. Please use relational field on message in the context of this PR, it's more robust, more declarative, and doesn't pollute master code with bad stuffs.

Reply:
I'll see what I can do for `lastSeenByCurrentPartnerMessageId` (new field), but I'll not change the existing field `lastSeenByCurrentPartnerMessageId` (not safe in stable and that would introduce ever more diff).